components/camera/platforms/jetson/camera.go

```diff
@@ -0,0 +1,114 @@
+package jetsoncamera
+
+// #include <sys/utsname.h>
+// #include <string.h>
+import "C"
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"runtime"
+	"unsafe"
+)
+
+// GetOSInformation pulls relevant OS attributes as an OSInformation struct
+// Kernel and Device will be "unkown" if unable to retrieve info from the filesystem
+// returns an error if kernel version or device name is unavailable
+func DetectOSInformation() (OSInformation, error) {
+	kernelVersion, err := getKernelVersion()
+	if err != nil {
+		return OSInformation{}, fmt.Errorf("failed to get kernel version: %w", err)
+	}
+	deviceName, err := getDeviceName()
+	if err != nil {
+		return OSInformation{}, fmt.Errorf("failed to get device name: %w", err)
+	}
+	osInfo := OSInformation{
+		Name:   runtime.GOOS,
+		Arch:   runtime.GOARCH,
+		Kernel: kernelVersion,
+		Device: deviceName,
+	}
+	return osInfo, nil
+}
+
+// getKernelVersion returns the Linux kernel version ($ uname -r)
+func getKernelVersion() (string, error) {
+	var utsName C.struct_utsname
+	if C.uname(&utsName) == -1 {
+		return Unknown, fmt.Errorf("uname information unavailable (%v)", utsName)
+	}
+	release := C.GoString((*C.char)(unsafe.Pointer(&utsName.release[0])))
+	return release, nil
+}
+
+// getDeviceName returns the model name of the device
+// ($ cat /sys/firmware/devicetree/base/model)
+func getDeviceName() (string, error) {
+	devicePath := "/sys/firmware/devicetree/base/model"
+	device, err := os.ReadFile(devicePath)
+	if err != nil {
+		return Unknown, err
+	}
+	return string(bytes.TrimRight(device, "\x00")), nil
+}
+
+// Validate checks if the daughterboard and driver are supported and installed on the device
+func Validate(osInfo OSInformation, daughterboardName string, driverName string) error {
+	board, ok := cameraInfoMappings[osInfo.Device]
+	if !ok {
+		return fmt.Errorf("the %s device is not supported on this platform", osInfo.Device)
+	}
+	daughterboard, ok := board.Daughterboards[daughterboardName]
+	if !ok {
+		return fmt.Errorf("the %s daughterboard is not supported on this platform", daughterboardName)
+	}
+	driver, ok := board.Modules[driverName]
+	if !ok {
+		return fmt.Errorf("the %s driver is not supported on this platform", driverName)
+	}
+	err := checkDaughterBoardConnected(daughterboard)
+	if err != nil {
+		return fmt.Errorf("the %s daughterboard is not connected or not powerd on. Please check daughter-board conenction to the %s", daughterboardName, osInfo.Device)
+	}
+	err = checkDriverInstalled(osInfo.Kernel, driver)
+	if err != nil {
+		return fmt.Errorf("the %s driver not installed. Please follow instructions for driver installation", driverName)
+	}
+
+	return fmt.Errorf("the %s daughterboard is connected and %s camera driver is installed on the %s. please check that the video path is correct and driver is working", daughterboardName, driverName, osInfo.Device)
```

Comment:
From @penguinland 

...it looks to me like this function is guaranteed to return a non-nil error. If everything is assembled/configured properly, what does the happy path through this function look like? Aren't we going to have errors no matter what?

Reply:
Yes this is a good point, there is no happy path...

`Validate` only gets called if there was an error getting the `cameraSource` in `webcam.go`, so we assume something is already wrong.

Reply:
Then I'd rename this, unless `Validate` always has that semantic meaning for cameras. For non-camera components, `Validate` checks that the config file looks plausible, and it normally returns `nil`. Maybe rename to `FindError` or `DetectError` or `DiagnoseProblem` or something?

Reply:
agreed - changed to `DetectError`